### PR TITLE
sign: Expose the Protected header

### PR DESCRIPTION
**Pull request by puiterwijk**
This adds two new functions to `COSESign1`:
- `new_with_protected`, to allow users to pass in a protected header
- `get_protected_and_payload`, to allow users to retrieve the protected
  header

These two functions allow users to use the protected header, in case
their use of COSE uses values in this header.

Signed-off-by: Patrick Uiterwijk <patrick@puiterwijk.org>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the Apache License Version 2.0, as specified in the `LICENSE` file of this repository.

### CODE REVIEW

**Comment by puiterwijk** [comment]
I don't think the 1.38 failure is due to me, that is because of something in `serde_with_macros` (pulled in by `serde_with`).

**Comment by petreeftime** [comment]
Seems the serde_with 1.5 crate has a new version. I'll see if I can bump the minimum rust version or pin the crate version more specifically.

**Comment by petreeftime** [comment]
https://github.com/awslabs/aws-nitro-enclaves-cose/pull/9 should fix the CI issue.

File: src/sign.rs [hunk 1]
```diff
@@ -773,6 +801,36 @@ mod tests {
...
+        let cose_doc2 = COSESign1::from_bytes(&cose_doc1.as_bytes(false).unwrap()).unwrap();
+
+        let (protected, payload) = cose_doc2
+            .get_protected_and_payload(Some(&ec_public))
+            .unwrap();
+
+        assert_eq!(
+            protected.get(&CborValue::Integer(1)),
+            Some(&CborValue::Integer(-7)),
+        );
```
Comment (petreeftime) [inline src/sign.rs:826]: This behavior is slightly unexpected.

I think `new_with_protected` should either:
* copy the protected header as is and not add or remove anything from it
* add additional fields into protected, but only if there is no conflict; fail if there is a conflict
* copy the protected header as is, but verify the variables it sets and fail if it thinks there's an error

I like the first option better, and `new` should just set the protected fields as it did until now.

**Review by petreeftime — CHANGES_REQUESTED** [review]
This is a very useful change, thank you! I left some comments about default behavior of `new_with_protected` inline.

File: src/sign.rs (hunk 1 above)
Comment (puiterwijk) [reply src/sign.rs:826]: Thanks for the comments. I implemented the first option.

**Review by petreeftime — APPROVED** [review]

**Review by andraprs — APPROVED** [review]

**Pull request merged**